docs: kubernetes operator madr

shaneutt opened the pull request:
The purpose of this PR is to add a first draft of an MADR that proposes the creation of a Kubernetes operator for Kuma mesh.

I have intentionally filled out only the context section for at least now to get the conversation started. Please feel free to point PRs at this PR to make changes: I'm certain I'm not the only one interested in this topic, so I would like to invite other people to help me co-author this MADR.

jakubdyszkiewicz commented:
For me, the biggest reason to introduce an operator is to simplify upgrades, which is the biggest operational pain point of the mesh. I think I'd focus on this use case taking into account multizone.

Relevant docs https://kuma.io/docs/2.0.x/documentation/upgrades/

shaneutt commented:
Thanks for the feedback @jakubdyszkiewicz and @lahabana.

> For me, the biggest reason to introduce an operator is to simplify upgrades

In the first draft I was definitely aware of the fact that upgrades are the big draw here and so it was the first example of a deliverable I wrote in the context section. Given the feedback however, I assumed you were hoping this would be emphasized more strongly and explicitly, as such I added it to the `Decision Drivers` in 468a2c4c72d10c42a956326b3144d2d034c97aaa. Let me know if there's more emphasis you feel needs to be made.

> While a single zone operator would still be useful (.e.g help with upgrading sidecars for example) the real value would be with multi-zone

I take this to specifically mean multi-cluster is a `Decision Driver`, as such I've added it to that section and started filling it out as one of the considered options in bb81006cdf7d745dbc4a8d53cae23225535c22b5. Let me know if you feel this represents your interests satisfactorily.

shaneutt commented:
I would also like to get some review from @doctorwu on this as he's an expert on Kuma operations and I am aware of him having his own operator-esque management tool that he developed and actively uses for Kuma, so if there's any insights from that project that can help us in fleshing out this MADR it would be greatly appreciated :bow:

slonka commented:
Pinging again @doctorwu for review.

doctorwu commented on docs/madr/decisions/011-mesh-operator.md:
```diff
@@ -0,0 +1,100 @@
+* *Option 2*: Build a Kuma operator with multi-cluster/multi-zone lifecycle
+  management built in.
+
+## Decision Outcome
```
From the enterprise perspective, having the ability to install a multi-zone mesh across multiple clusters from a single operator would be useful. In most of my engagements with customers, standalone mode is not used. Multi-zone is the only deployment option for large organizations.

shaneutt replied:
Sorry, this is just copy/paste from the template. We have not chosen any specific option yet. I can probably just remove all the templated text below if that's more clear.

shaneutt commented:
What is the criteria for merging a `proposed` MADR in this project?

slonka commented:
> What is the criteria for merging a proposed MADR in this project?

It's 3 approvals https://github.com/kumahq/kuma/blame/e912d2112d7eb5350d884dbf5bd3e365db653cac/docs/madr/decisions/001-adopt-madr.md#L59, you can read the whole thing as there is more info about the process there.

It's been open for a while and it feels like it's a bit stuck. Would be great to hear from the people who already looked at this PR on what's missing to get the approves ( @jakubdyszkiewicz @lahabana @doctorwu ). From my side: the template stuff needs to be filled out or removed if it does not apply. Also is it possible to add more detail on the implementation? I've never written an operator - if it's really straightforward then maybe it's not needed - if it's more complicated maybe there are some aspects that need to be hashed out before implementing it?

jakubdyszkiewicz commented:
What I see here are a problem statement and two very high-level options. I'd love to see what would each option mean.
For example "manage the lifecycle of the Kuma control plane and all required subcomponents". What exactly would happen in case of upgrade. We also specify user-facing UX in the MADRs. Stuff like
1) What would operator CRD look like
2) How does it look like to install/upgrade Kuma using the operator?

+1 to template stuff mentioned above.

I'm excited to see the initial effort, but I think we need to have a bit more so the MADR can be accepted.

lahabana commented:
> What is the criteria for merging a proposed MADR in this project?

We don't really do "proposed" MADR. The conversation just fully happens on the PR and then get merged when approved. This usually involves doing some renumbering at some point but it avoids having 10s of in-progress never finished MADRs

shaneutt commented:
> the template stuff needs to be filled out or removed if it does not apply

> +1 to template stuff mentioned above.

Sounds good, this is now removed in 0f661146ead228dc6f102ccf7543de2019438777

> also is it possible to add more detail on the implementation? I've never written an operator - if it's really straightforward then maybe it's not needed - if it's more complicated maybe there are some aspects that need to be hashed out before implementing it?

> I'm excited to see the initial effort, but I think we need to have a bit more so the MADR can be accepted.

> We don't really do "proposed" MADR. The conversation just fully happens on the PR and then get merged when approved. This usually involves doing some renumbering at some point but it avoids having 10s of in-progress never finished MADRs

It was an explicit intention of mine to not include implementation details for the first pass: my thinking was that it was valuable to capture the **what** and the **why** first to make sure we're all aligned, and then later start adding the **how**. Also I was hoping to get someone to collaborate with me on this for the "how", e.g. do some brainstorming e.t.c. and then in a follow-up PR add that. Sorry for my misconceptions, given the desire for a more complete MADR up front I'm going to close this for now until I have someone who wants to collaborate and we'll fill in some of the deeper details and re-open. Thank you!

shaneutt commented:
> the template stuff needs to be filled out or removed if it does not apply

> +1 to template stuff mentioned above.

Sounds good, this is now removed in 0f661146ead228dc6f102ccf7543de2019438777

> also is it possible to add more detail on the implementation? I've never written an operator - if it's really straightforward then maybe it's not needed - if it's more complicated maybe there are some aspects that need to be hashed out before implementing it?

> I'm excited to see the initial effort, but I think we need to have a bit more so the MADR can be accepted.

> We don't really do "proposed" MADR. The conversation just fully happens on the PR and then get merged when approved. This usually involves doing some renumbering at some point but it avoids having 10s of in-progress never finished MADRs

It was an explicit intention of mine to not include implementation details for the first pass: my thinking was that it was valuable to capture the **what** and the **why** first to make sure we're all aligned, and then later start adding the **how**. Also I was hoping to get someone to collaborate with me on this for the "how", e.g. do some brainstorming e.t.c. and then in a follow-up PR add that. Sorry for my misconceptions, given the desire for a more complete MADR up front I'm going to close this for now until I have someone who wants to collaborate and we'll fill in some of the deeper details and re-open. Thank you!
